website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+    if cursor.fetchone()[0] == 0:
+        products = []
+        for i in range(1, 51):
+            products.append((
+                i,
+                f'Product {i}',
+                round(10 + i * 0.5, 2)
+            ))
+        cursor.executemany('''
+        INSERT INTO Products (ProductID, ProductName, Price)
+        VALUES (?, ?, ?)
+        ''', products)
+
+    # Populate Orders table if empty
+    cursor.execute('SELECT COUNT(*) FROM Orders')
+    if cursor.fetchone()[0] == 0:
+        orders = []
+        from datetime import datetime, timedelta
+        base_date = datetime(2023, 1, 1)
+        for i in range(1, 51):
+            order_date = base_date + timedelta(days=i)
+            orders.append((
+                i,
+                i % 50 + 1,  # CustomerID between 1 and 50
+                order_date.strftime('%Y-%m-%d')
+            ))
+        cursor.executemany('''
+        INSERT INTO Orders (OrderID, CustomerID, OrderDate)
+        VALUES (?, ?, ?)
+        ''', orders)
+
+    # Populate OrderDetails table if empty
+    cursor.execute('SELECT COUNT(*) FROM OrderDetails')
+    if cursor.fetchone()[0] == 0:
+        order_details = []
+        for i in range(1, 51):
+            order_details.append((
+                i,
+                i % 50 + 1,  # OrderID between 1 and 50
+                i % 50 + 1,  # ProductID between 1 and 50
+                (i % 5 + 1) * 2  # Quantity between 2 and 10
+            ))
+        cursor.executemany('''
+        INSERT INTO OrderDetails (OrderDetailID, OrderID, ProductID, Quantity)
+        VALUES (?, ?, ?, ?)
+        ''', order_details)
+
+    conn.commit()
+
+def setup_database():
+    """Setup the database and return the connection."""
+    db_file = 'data/database.db'
+    if not os.path.exists('data'):
+        os.makedirs('data')
+
+    db_exists = os.path.exists(db_file)
+
+    conn = create_connection(db_file)
+
+    if not db_exists:
+        print("Setting up the database...")
+        create_tables(conn)
+        populate_tables(conn)
+    else:
+        print("Database already exists. Skipping setup.")
+
+    return conn
+```
+
+### SQL Generation Chain
+
+The **SQL Generation Chain** (`sql_gen_chain`) is the backbone of automated SQL query generation in our workflow. This chain leverages LangChain's modular capabilities and OpenAI's advanced natural language processing to transform user questions into precise and executable SQL queries.
+
+**Core Features**:
+
+- **Prompt-Driven Generation**: Begins with a thoughtfully designed prompt that integrates the database schema and documentation snippets, ensuring queries are contextually accurate.
+
+- **Structured Responses**: Delivers outputs in a predefined format, including:
+
+- A **description** of the query's purpose.
+
+- The corresponding **SQL code** ready for execution.
+
+- **Adaptable and Reliable**: Uses GPT-4 for robust, consistent query generation, minimizing manual effort and errors.
```

Comment:
Should we update this to use `gpt-4o-mini`? The tool calling functionality with that LLM build is far superior to base gpt-4

Reply:
Great suggestion! I’ve updated the workflow to use gpt-4o-mini.